### PR TITLE
Feature message queue

### PR DESCRIPTION
**Pull request by mattkasun**
closing #621
closing #514

### CODE REVIEW

File: controllers/node.go [hunk 1]
```diff
@@ -427,6 +429,15 @@ func uncordonNode(w http.ResponseWriter, r *http.Request) {
 	if err = runServerPeerUpdate(node.Network, false); err != nil {
 		logger.Log(1, "internal error when approving node:", nodeid)
 	}
+	go func() {
+		if err := mq.NodeUpdate(&node); err != nil {
+			logger.Log(1, "error publishing node update"+err.Error())
+		}
+		if err := mq.UpdatePeers(&node); err != nil {
+			logger.Log(1, "error publishing peer update "+err.Error())
+			return
```
Comment (0xdcarns) [inline controllers/node.go:438]: empty return?

File: controllers/node.go [hunk 2]
```diff
@@ -598,7 +652,12 @@ func deleteNode(w http.ResponseWriter, r *http.Request) {
 		returnErrorResponse(w, r, formatError(err, "internal"))
 		return
 	}
-
+	node.Action = models.NODE_DELETE
+	go func() {
+		if err := mq.NodeUpdate(&node); err != nil {
+			logger.Log(1, "error publishing node delete "+err.Error())
```
Comment (0xdcarns) [inline controllers/node.go:658]: no need to concatenate strings, just do "error publishing node delete", err.Error(). The logger will join with spaces.

File: logic/peers.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitl/netmaker/models"
+	"github.com/gravitl/netmaker/netclient/ncutils"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+func GetPeerUpdate(node *models.Node) (models.PeerUpdate, error) {
```
Comment (0xdcarns) [inline logic/peers.go:15]: Add comment for public function

File: models/mqtt.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package models
+
+import "golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+type PeerUpdate struct {
```
Comment (0xdcarns) [inline models/mqtt.go:5]: comments required for public types, funcs and variables

File: mq/mq.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/gravitl/netmaker/database"
+	"github.com/gravitl/netmaker/logger"
+	"github.com/gravitl/netmaker/logic"
+	"github.com/gravitl/netmaker/models"
+	"github.com/gravitl/netmaker/servercfg"
+)
+
+// default message handler - only called in GetDebug == true
```
Comment (0xdcarns) [inline mq/mq.go:17]: malformed comment

File: netclient/daemon/systemd.go [hunk 6]
```diff
@@ -29,49 +29,28 @@ func SetupSystemDDaemon(interval string) error {
...
 	}
-
-	if !ncutils.FileExists("/usr/local/bin/netclient") {
-		os.Symlink("/etc/netclient/netclient", "/usr/local/bin/netclient")
-	}
-	if !ncutils.FileExists("/etc/netclient/netclient") {
-		err = ncutils.Copy(binarypath, "/etc/netclient/netclient")
+	//install binary
+	//should check if the existing binary is the corect version -- for now only copy if file doesn't exist
+	if !ncutils.FileExists("/usr/sbin/netclient") {
```
Comment (0xdcarns) [inline netclient/daemon/systemd.go:34]: "/usr/sbin/" should probably be a constant

File: netclient/daemon/systemd.go [hunk 7]
```diff
@@ -80,25 +59,17 @@ WantedBy=timers.target
...
-			log.Println(err)
-			return err
-		}
-	}
-
 	_, _ = ncutils.RunCmd("systemctl enable netclient.service", true)
 	_, _ = ncutils.RunCmd("systemctl daemon-reload", true)
-	_, _ = ncutils.RunCmd("systemctl enable netclient.timer", true)
-	_, _ = ncutils.RunCmd("systemctl start netclient.timer", true)
+	_, _ = ncutils.RunCmd("systemctl start netclient.server", true)
```
Comment (0xdcarns) [inline netclient/daemon/systemd.go:64]: I think this should be netclient.service?

File: netclient/functions/daemon.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+			_, err := ncutils.RunCmd("/usr/bin/resolvectl revert "+cfg.Node.Interface, true)
+			if err != nil {
+				ncutils.Log("error applying dns" + err.Error())
+			}
+		}
+	}()
+}
+
+// UpdatePeers -- mqtt message handler for /update/peers/<NodeID> topic
+var UpdatePeers mqtt.MessageHandler = func(client mqtt.Client, msg mqtt.Message) {
```
Comment (0xdcarns) [inline netclient/functions/daemon.go:172]: I think almost all of these Update.. funcs can be private

File: controllers/node.go (hunk 1 above)
Comment (mattkasun) [reply controllers/node.go]: removed this and others

File: controllers/node.go (hunk 2 above)
Comment (mattkasun) [reply controllers/node.go]: updated this and other logger.Log calls

File: models/mqtt.go (hunk 4 above)
Comment (mattkasun) [reply models/mqtt.go:5]: added comments

File: netclient/daemon/systemd.go (hunk 6 above)
Comment (mattkasun) [reply netclient/daemon/systemd.go]: added constant... also updated netclient/daemon/macos.go

File: netclient/daemon/systemd.go (hunk 7 above)
Comment (mattkasun) [reply netclient/daemon/systemd.go]: probably why it wasn't starting :1st_place_medal:

**Pull request merged**